### ISSUE TITLE
Consider creating an account after payment is successful.

### ISSUE THREAD

**Issue by senadir** [woothemes/woocommerce#50281]
### Describe the solution you'd like

Accounts are now created when you place an order, but payment can fail, but the customer still has an account created. This can pollute the user table and cause confusion for future visits (as they can't place an order because we tell them an account already exists).

This also might conflict with the follow up account creation and enhancements we're doing (for example, customers can't see the notice that tells them we sent an email with password).

Additionally, some shops might only allow certain access to things for logged in users, assuming they're paid users. You can place a failed order and still get access.

**Comment by senadir** [comment]
cc @mikejolley

**Comment by mikejolley** [comment]
@senadir 

Create account first:
+ Any issues with account creation (password, existing) are flagged
- Order may fail, pollutes table, may cause errors

Create order first:
+ Non-blocking payment
- Account creation errors unnoticed
- Order may not be associated with account

Also need to consider when checkout requires an account.

**Comment by senadir** [comment]
> Also need to consider when checkout requires an account.

Good point, if account creation fails, we need to figure out what happens, do we keep payment but fail the order?

**Comment by senadir** [comment]
I think what I had in mind is that we switch up account creation call and payment processing, but everything stays the same, if account creation fails, we stop Checkout, but we run into issues like redirect payments, holding and releasing payment...

**Comment by mikejolley** [comment]
We could add a pre-payment validation for username + email requirements, but not create account until afterwards. But this will require an extra call to update the order and link it to the new account.

**Comment by pierorocca** [comment]
Subscriptions use case is imo the one that would guide the right order of events because the account creation is mandatory as is storing the credit card details for future subscriptions payments. 

Creating the account after the payments approval provides some level of assurance, not fully, that the card and the shopper are legitimate because it goes through more risk screening across the payments ecosystem. If the payment declined or is blocked for risk, no account should get created and the card is definitely not stored. WooPay for example does not create user accounts if payment fails and "save my info" was checked. We'd have a lot of useless accounts with no saved payment method.

If the payment is approved but account creation fails and the card can't be stored, then depending if an account is mandatory or not, there needs to be a recovery flow. Either a way to retry creating an account or linking an email to an order (@mikejolley has an enhancement idea for that) or the order is reversed and the shopper is refunded.

**Comment by pierorocca** [comment]
The other option, and thinking about it is the best option, is to get rid of the account creation IN the checkout. Either the shopper clicks Login and creates an account and is then redirected back to checkout or they create an account post order. That way we don't have to deal with this question.

Do you know off the top of your head if delayed account creation is on but a shopper purchases a subscription and an account is created and payment method stores, does the delayed account creation UI display on the order confirmation page?

**Comment by mikejolley** [comment]
> Do you know off the top of your head if delayed account creation is on but a shopper purchases a subscription and an account is created and payment method stores, does the delayed account creation UI display on the order confirmation page?

Subscriptions forces registration as its a requirement for checkout (they do it via hooks and manipulating settings). In that case, delayed account won't work. So we have to either leave it as an option during checkout, or work out a decent pre-checkout registration system as an alternative for the use cases that require it.

**Comment by pierorocca** [comment]
Thanks @mikejolley. I did some testing and see that delayed account creation is not shown, expected, since the account is already created. Sorry I should have just tried to confirm the use case had been accounted for.

**Comment by pierorocca** [comment]
Hi @senadir, 

I like your idea of preventing account creation for failed payments. For approved payments that fail account creation, rare, a recovery path is needed especially for a subscription. Subscriptions I'm recommending 1-4 below. For non-subs, 1-2 is the minimum. 3 is nice to have. 4 is not needed. 

Thoughts?

1. Immediate retry (synchronous, same request)

Before giving up, retry 2 times in the same request. Handles the most common real-world cause — a transient DB hiccup. If it succeeds on retry, the shopper never sees anything wrong.

2. Background retry via Action Scheduler

WooCommerce ships with Action Scheduler. If synchronous retries exhaust, schedule a deferred job to retry account creation and link the order. Order goes to on-hold in the interim. Action Scheduler already handles retries with backoff natively.

as_schedule_single_action(
    time() + 60,
    'wc_retry_account_creation_for_order',
    [ 'order_id' => $order_id, 'email' => $email, 'attempt' => 1 ]
);
$order->update_status( 'on-hold', 'Account creation failed, retrying.' );

3. Tokenized email to the shopper

If background retries also exhaust, send the shopper an email to their provided address with a tokenized link: "Your payment was successful. Click here to activate your account and access your subscription." The link pre-fills their details, creates the account, and links the pending order to it.

WooCommerce already has infrastructure for this — the delayed account creation email flow uses a similar pattern with a nonce tied to the order.

4. Admin order note

Alongside option 3, surface an admin order note and place the order on-hold so the merchant knows an order is stuck. They can manually trigger account creation or issue a refund if the customer never responds.

The natural cascade is: retry synchronously → retry via Action Scheduler → tokenized email + admin notice → manual intervention. The first two levels handle infrastructure failures silently. The email handles the case where retries keep failing (a persistent issue). Admin order note and on-hold status is the last resort.

**Comment by senadir** [comment]
I think there is value in giving this another chance, but this is an enhancement that should sit a bit low in the backlog, tbh
